### ISSUE TITLE
Xcode 16 beta issues

### ISSUE THREAD

**Issue by mapierce** [MakeAWishFoundation/SwiftyMocky#362]
Generating mocks with version `4.1.0` of SwiftyMocky results in odd behaviour in my test target when using Xcode 16 beta (any version). I get a lot of errors like:
```
Type 'Any' cannot conform to 'ExpressiblyByArrayLiteral'
Type 'Any' cannot conform to 'ExpressiblyByStringLiteral'
Type 'Any' cannot conform to 'ExpressiblyByIntLiteral'
```
around lines such as:
```swift
dictionary["key"] = value
```
where `dictionary` is defined as type `[String: Any]`.

Removing the `Mock.generated.swift` file from the "Compile Sources" step of the build phases gets rid of these errors and allows it to build (where SwiftMocky isn't used and this the mock file is not depended on).

I attempted to update to `4.2.0` of SwiftyMocky, but running `generate` results in a large amount of warnings, a sample of which look like:
```
warning: Failed to unarchive cache for SomeFile1.swift due to error, re-parsing file
warning: Failed to unarchive cache for SomeFile2.swift due to error, re-parsing file
warning: Failed to unarchive cache for SomeFile3.swift due to error, re-parsing file
warning: Failed to unarchive cache for SomeFile4.swift due to error, re-parsing file
warning: Failed to unarchive cache for SomeFile5.swift due to error, re-parsing file
warning: Failed to unarchive cache for SomeFile6.swift due to error, re-parsing file
warning: Failed to unarchive cache for SomeFile7.swift due to error, re-parsing file
warning: Failed to unarchive cache for SomeFile8.swift due to error, re-parsing file
warning: Failed to unarchive cache for SomeFile9.swift due to error, re-parsing file
```
but then I also get what appear to be Swift 6 concurrency errors like:
```
/private/var/folders/64/sqh1_t1914v44d7tpgy0cft40000gn/T/SwiftTemplate/1.8.0/Sources/SourceryRuntime/Array+Parallel.swift:21:17: error: mutable capture of 'inout' parameter 'buffer' is not allowed in concurrently-executing code
19 |         return result.withUnsafeMutableBufferPointer { buffer in
20 |             DispatchQueue.concurrentPerform(iterations: buffer.count) { idx in
21 |                 buffer[idx] = transform(self[idx])
   |                 `- error: mutable capture of 'inout' parameter 'buffer' is not allowed in concurrently-executing code
22 |             }
23 |             return buffer.map { $0! }
[42/50] Compiling SourceryRuntime Annotations.swift
/private/var/folders/64/sqh1_t1914v44d7tpgy0cft40000gn/T/SwiftTemplate/1.8.0/Sources/SourceryRuntime/Array+Parallel.swift:21:17: error: mutable capture of 'inout' parameter 'buffer' is not allowed in concurrently-executing code
19 |         return result.withUnsafeMutableBufferPointer { buffer in
20 |             DispatchQueue.concurrentPerform(iterations: buffer.count) { idx in
21 |                 buffer[idx] = transform(self[idx])
   |                 `- error: mutable capture of 'inout' parameter 'buffer' is not allowed in concurrently-executing code
22 |             }
23 |             return buffer.map { $0! }
[43/50] Compiling SourceryRuntime Array+Parallel.swift
/private/var/folders/64/sqh1_t1914v44d7tpgy0cft40000gn/T/SwiftTemplate/1.8.0/Sources/SourceryRuntime/Array+Parallel.swift:21:17: error: mutable capture of 'inout' parameter 'buffer' is not allowed in concurrently-executing code
19 |         return result.withUnsafeMutableBufferPointer { buffer in
20 |             DispatchQueue.concurrentPerform(iterations: buffer.count) { idx in
21 |                 buffer[idx] = transform(self[idx])
   |                 `- error: mutable capture of 'inout' parameter 'buffer' is not allowed in concurrently-executing code
22 |             }
23 |             return buffer.map { $0! }
[44/50] Compiling SourceryRuntime Array.swift
/private/var/folders/64/sqh1_t1914v44d7tpgy0cft40000gn/T/SwiftTemplate/1.8.0/Sources/SourceryRuntime/Array+Parallel.swift:21:17: error: mutable capture of 'inout' parameter 'buffer' is not allowed in concurrently-executing code
19 |         return result.withUnsafeMutableBufferPointer { buffer in
20 |             DispatchQueue.concurrentPerform(iterations: buffer.count) { idx in
21 |                 buffer[idx] = transform(self[idx])
   |                 `- error: mutable capture of 'inout' parameter 'buffer' is not allowed in concurrently-executing code
22 |             }
23 |             return buffer.map { $0! }
```
So I appear to be fully blocked from running any of my tests on Xcode 16. 

Hoping there's a fix for this soon 🙏 as it's a pretty big blocker.

**Update**

I dug into this a bit and it seems to be related to [this issue](https://stackoverflow.com/questions/78612277/xcode-16-warning-extension-declares-a-conformance-of-imported-type-this-wil) and [SE-0364](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0364-retroactive-conformance-warning.md). It looks like it's occurring as a result of the extensions on the `Optional` type to add conformance to expressible by string / integer / boolean etc. literal. It looks like the best option in a scenario like this is to create a new wrapper type for optionals instead of explicitly extending the Optional type to work with `Parameter`. I would suggest something like `OptionalParameter` which would change the mocked method signature of a function like:
```swift
func fetchResults(for searchTerm: String, source: String?)
```
from this:
```swift
public static func fetchResults(for searchTerm: Parameter<String>, source: Parameter<String?>) -> Perform {
    return Perform(method: m_fetchResults__forSearchTermsource, performs: perform)
}
```
to something like:
```swift
public static func fetchResults(for searchTerm: Parameter<String>, source: OptionalParameter<String>) -> Perform {
    return Perform(method: m_fetchResults__forSearchTermsource, performs: perform)
}
```

**Comment by kyzmitch** [comment]
similar here:
```
❌  error: SWIFT_VERSION '6.0' is unsupported, supported versions are: 4.0, 4.2, 5.0. (in target 'MyTarget' from project 'myProject')
```
even if I select Xcode 15.4 which is supported, I guess somehow need to set Swift version to 5.x in Xcode project for the unit tests, but what about the actual module which I want to use 6.0 IDK

**Comment by bukira** [comment]
@mapierce is this going to be point into a release so I can use it with the CLI, still getting the Xcode 16 issues using cli 4..20 and swiftymocky 4.2.0

**Comment by mapierce** [comment]
Hey @bukira, I'm not a maintainer on SwiftyMocky, so I don't have control over their release process. To work around that right now I point to master of this repo in my Podfile instead of a specific version.

**Comment by bukira** [comment]
@mapierce ah right thxs, I use SPM not cocpods :-( does this make swiftymock compatible with swift 6 concurrency?

**Comment by mapierce** [comment]
@bukira just point the package to master branch instead of a version

**Comment by bukira** [comment]
@mapierce I did but when I use the CLI I am still 4.20? I am still getting errors

❌  Error: ShellOut encountered an error
Status code: 1
Message: "🌱 Encountered error during "swift build -c release --product sourcery -Xswiftc -target -Xswiftc arm64-apple-macosx14.6". Use --verbose to see full output
🌱  Failed to build Sourcery 1.8.0 with SPM"
Output: "🌱 Cloning Sourcery 1.8.0
🌱 Resolving package
🌱 Building product sourcery
Building for production...
[0/31] Write sources
[4/31] Copying lib_InternalSwiftSyntaxParser.dylib
[4/31] Write sources
[8/31] Copying ejs.js
[8/31] Write sources
[8/31] Copying lib_InternalSwiftSyntaxParser.framework
[8/31] Write sources
[20/31] Write swift-version--58304C5D6DBC2206.txt
[21/33] Compiling atomic-counter.c
[22/34] Compiling writer.c
[23/34] Compiling reader.c
[24/34] Compiling api.c
[25/34] Compiling parser.c
[26/36] Compiling emitter.c
[27/36] Compiling scanner.c
[28/37] Compiling TryCatch TryCatch.m
[30/37] Compiling AEXML Document.swift
[31/37] Compiling PathKit PathKit.swift
[32/41] Compiling SourceryJS EJSTemplate.swift
[33/41] Compiling SourceryUtils Path+Extensions.swift
[34/41] Compiling Stencil Context.swift
[35/41] Compiling Yams Constructor.swift
[36/42] Compiling Commander ArgumentConvertible.swift
[37/42] Compiling StencilSwiftKit CallMacroNodes.swift
[38/42] Compiling SourceryRuntime AccessLevel.swift
/private/var/folders/ww/hwd5bq0s6cddnvr08x3jdspr0000gn/T/mint/github.com_krzysztofzablocki_Sourcery/SourceryRuntime/Sources/Array+Parallel.swift:21:17: warning: capture of 'buffer' with non-sendable type 'UnsafeMutableBufferPointer<T?>' in a `@Sendable` closure; this is an error in the Swift 6 language mode
19 |         return result.withUnsafeMutableBufferPointer { buffer in
20 |             DispatchQueue.concurrentPerform(iterations: buffer.count) { idx in
21 |                 buffer[idx] = transform(self[idx])
   |                 `- warning: capture of 'buffer' with non-sendable type 'UnsafeMutableBufferPointer<T?>' in a `@Sendable` closure; this is an error in the Swift 6 language mode
22 |             }
23 |             return buffer.map { $0! }

Swift.UnsafeMutableBufferPointer:1:23: note: generic struct 'UnsafeMutableBufferPointer' does not conform to the 'Sendable' protocol
1 | @frozen public struct UnsafeMutableBufferPointer<Element> : Copyable where Element : ~Copyable {
  |                       `- note: generic struct 'UnsafeMutableBufferPointer' does not conform to the 'Sendable' protocol
2 |     public let count: Int
3 |     @inlinable public init(start: UnsafeMutablePointer<Element>?, count: Int)

/private/var/folders/ww/hwd5bq0s6cddnvr08x3jdspr0000gn/T/mint/github.com_krzysztofzablocki_Sourcery/SourceryRuntime/Sources/Composer.swift:340:13: warning: initialization of immutable value 'modules' was never used; consider replacing with assignment to '_' or removing it
338 | 
339 |         let unique = state.typeMap
340 |         let modules = state.modules
    |             `- warning: initialization of immutable value 'modules' was never used; consider replacing with assignment to '_' or removing it
341 |         let typealiases = state.resolvedTypealiases
342 | 

/private/var/folders/ww/hwd5bq0s6cddnvr08x3jdspr0000gn/T/mint/github.com_krzysztofzablocki_Sourcery/SourceryRuntime/Sources/Composer.swift:341:13: warning: initialization of immutable value 'typealiases' was never used; consider replacing with assignment to '_' or removing it
339 |         let unique = state.typeMap
340 |         let modules = state.modules
341 |         let typealiases = state.resolvedTypealiases
    |             `- warning: initialization of immutable value 'typealiases' was never used; consider replacing with assignment to '_' or removing it
342 | 
343 |         if let name = typeName.actualTypeName {

/private/var/folders/ww/hwd5bq0s6cddnvr08x3jdspr0000gn/T/mint/github.com_krzysztofzablocki_Sourcery/SourceryRuntime/Sources/Log.swift:60:1: warning: extension declares a conformance of imported type 'String' to imported protocol 'Error'; this will not behave correctly if the owners of 'Swift' introduce this conformance in the future
58 | }
59 | 
60 | extension String: Error {}
   | |- warning: extension declares a conformance of imported type 'String' to imported protocol 'Error'; this will not behave correctly if the owners of 'Swift' introduce this conformance in the future
   | `- note: add '@retroactive' to silence this warning
61 |

**Comment by mapierce** [comment]
That's an issue with Sourcery, not SwiftyMocky. I have a fix for that merged in on [Sourcery's repo](https://github.com/krzysztofzablocki/Sourcery/pull/1363) as well, but they have not run a release yet. You can point SwiftyMocky to use master version of Sourcery in your mockfile, or just download master of Sourcery if you're using the CLI independently.

**Comment by bukira** [comment]
@mapierce your a genius, so how would I do that then? either or both

**Comment by mapierce** [comment]
@bukira depends what you're doing. If you're not using Sourcery directly, just via SwiftyMocky, then you should update your Mockfile to use master of Sourcery. There's docs somewhere in this repo about that. 

Otherwise if you're using Sourcery directly then you should download that.

**Comment by bukira** [comment]
@mapierce thans for this, I got it to use master Sourcey in the Mockfile however I still get the same error repeating over about "characters"

error: Building for production...
[0/5] Write sources
[2/5] Write swift-version--58304C5D6DBC2206.txt
[4/7] Emitting module SourceryRuntime
[5/7] Compiling SourceryRuntime AccessLevel.swift
[6/9] Emitting module SwiftTemplate
[7/9] Compiling SwiftTemplate main.swift
/private/var/folders/ww/hwd5bq0s6cddnvr08x3jdspr0000gn/T/SwiftTemplate/196FF28E-DFE6-4A19-B0E9-44DAA76D89FA/2.2.5/Sources/SwiftTemplate/main.swift:251:96: error: 'characters' is unavailable: Please use String directly
 249 |     static func extractGenericsList(_ associatedTypes: [String]?) -> [String] {
 250 |         return associatedTypes?.flatMap {
 251 |             split($0, byFirstOccurenceOf: " where ").0.replacingOccurrences(of: " ", with: "").characters.split(separator: ":").map(String.init).first
     |                                                                                                `- error: 'characters' is unavailable: Please use String directly
 252 |         }.map { "\($0)" } ?? []
 253 |     }

Swift.String:5:16: note: 'characters' was obsoleted in Swift 5.0
3 |     public typealias CharacterView = String
4 |     @available(swift, deprecated: 3.2, obsoleted: 5.0, message: "Please use String directly")
5 |     public var characters: String { get set }
  |                `- note: 'characters' was obsoleted in Swift 5.0
6 |     @available(swift, deprecated: 3.2, obsoleted: 5.0, message: "Please mutate the String directly")
7 |     public mutating func withMutableCharacters<R>(_ body: (inout String) -> R) -> 


repeats 6 timers for different places in : 

/private/var/folders/ww/hwd5bq0s6cddnvr08x3jdspr0000gn/T/SwiftTemplate/196FF28E-DFE6-4A19-B0E9-44DAA76D89FA/2.2.5/Sources/SwiftTemplate/main.swift:

**Comment by bukira** [comment]
@mapierce how did you get this to work? doesnt work for me, using master of swiftymocky and master of sourcery ,I still get the issues about "characters"
